src/Features/Core/Portable/Completion/CompletionList.cs

```diff
@@ -53,19 +62,24 @@ public sealed class CompletionList
 
         private CompletionList(
             TextSpan defaultSpan,
-            ImmutableArray<CompletionItem> items,
+            IReadOnlyList<CompletionItem> items,
             CompletionRules? rules,
             CompletionItem? suggestionModeItem,
             bool isExclusive)
         {
             Span = defaultSpan;
+            ItemsInternal = items;
+
+            if (items is ImmutableArray<CompletionItem>)
+                _lazyItems = new(() => (ImmutableArray<CompletionItem>)ItemsInternal);
+            else
+                _lazyItems = new(() => ItemsInternal.ToImmutableArray(), System.Threading.LazyThreadSafetyMode.PublicationOnly);
```

Comment:
why PublicationOnly?

Reply:
Should I use `ExecutionAndPublication`? I don't think it matters much though, since the value is not accessed concurrently in Roslyn.

Reply:
well, i'm curious why this line is different than the one above.  :)  i would expect you'd have the same thread-safety-choices for both.

Note: you can also simplify this to a single lazy that does `ToImmutableArrayOrEmpty` inside (that extension already checks if the underlying type is an ImmutableArray).